include/visibility.h

```diff
@@ -13,18 +13,33 @@
 
 #if defined _WIN32 || defined __CYGWIN__
     #ifdef BUILDING_PHPCPP
-        #ifdef __GNUC__
-            #define PHPCPP_EXPORT __attribute__ ((dllexport))
-        #else
-            #define PHPCPP_EXPORT __declspec(dllexport) // Note: actually gcc seems to also supports this syntax.
+        #ifdef BUILDING_SHARED
+            #ifdef __GNUC__
+                #define PHPCPP_EXPORT __attribute__ ((dllexport))
+            #else
+                #define PHPCPP_EXPORT __declspec(dllexport) // Note: actually gcc seems to also supports this syntax.
+            #endif
+        #else /* build static library */
+          #define PHPCPP_EXPORT
         #endif
+        #define DLL_EXPORT
     #else
-        #ifdef __GNUC__
-            #define DLL_EXPORT __attribute__ ((dllimport))
-        #else
-            #define DLL_EXPORT __declspec(dllimport) // Note: actually gcc seems to also supports this syntax.
+        #ifdef BUILDING_SHARED
+            #ifdef __GNUC__
+                #define DLL_EXPORT __attribute__ ((dllimport))
+            #else
+                #define DLL_EXPORT __declspec(dllimport) // Note: actually gcc seems to also supports this syntax.
+            #endif
+        #else /* build static library */
+            #define DLL_EXPORT
         #endif
+        #define PHPCPP_EXPORT
     #endif
 #else
-    #define PHPCPP_EXPORT __attribute__ ((visibility ("default")))
+    #ifdef BUILDING_SHARED
+        #define PHPCPP_EXPORT __attribute__ ((visibility ("default")))
+    #else /* build static library */
+        #define PHPCPP_EXPORT
+    #endif
+    #define DLL_EXPORT
```

Comment:
This piece seems wrong. If the attribute is supported, DLL_EXPORT should be same as PHPCPP_EXPORT.

In general I see what you're striving to achieve, but it is not going to work well. An extension should define its own API inside it, using same specifier will still require additional compiler flags. Fe if there's a static phpcpp lib, get_module() inside the extension will still have to be DLL exported.

Thanks.